File: labs/module8/Fortran/base/dot_acc.f90
module dot_acc_m
contains
   subroutine dot_acc(A, B, C, m, n)
      implicit none
      real, intent(in) :: A(m,n)
      real, intent(in) :: B(m,n)
      real, intent(inout) :: C(n)
      integer, intent(in) :: m, n
    
      real :: temp
      integer :: i, j
    
!$acc parallel loop gang private(temp) copyin(A,B) copyout(C)
      do j = 1,n
         temp = 0.0
!$acc loop vector reduction(+:temp)
         do i = 1,m
            temp = temp + A(i,j) * B(i,j)
         end do
         C(j) = temp
      end do

      return
  end subroutine dot_acc
  
end module dot_acc_m

File: labs/module8/C/deviceptr/dot_acc.c
#include <stdio.h>
#include <stdlib.h>

void dot_acc(int *A, int *B, int *C, int m, int n)
{
	int temp;
#pragma acc parallel loop gang private(temp) \
 copyin(A[:m*n],B[:m*n]) copyout(C[:m])
	for( int i = 0; i < m; i++ ) {
		temp = 0;
#pragma acc loop vector reduction(+:temp)
		for( int j = 0; j < n; j++ ) {
			temp +=  A[i*n+j] * B[i*n+j];
		}
		C[i] = temp;
	}
}



